cli/demo,server: new auto-login URL path

knz opened the pull request:
Fixes #54877

Previous to this patch, the user of `cockroach demo` had to perform
two copy-paste operations from their terminal to use the admin UI: one
to copy the URL into their browser, and another oen to copy the
authentication password.

This was inconvenient and thus created friction to exploration and
bottom-up adoption.

This patch rectifies the situation by ensuring that the URL printed
int he terminal allows the user to access the UI directly, in just one
step.

![image](https://user-images.githubusercontent.com/642886/103227973-05d47680-4930-11eb-9ffd-94f553fb3750.png)

(see how the URL embeds the authentication credentials)

Technically, the approach prototyped here defines a new HTTP GET
endpoint `/demologin` which accepts the user/password parameters as
query arguments, with the same logic (more or less) as `/login`. This
endpoint is reported by (and used only by) `cockroach demo`.

Release note (cli change): The DB console URL printed by `cockroach
demo` now automatically logs the user in when pasted into a web browser.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56740)
<!-- Reviewable:end -->

knz commented:
NB: wikipedia is telling me that support for username:pw in the URL before the hostname, separated with `@`, is "deprecated". I'm not sure if that's really true?
@dhartunian @bdarnell might know more about it?

bdarnell commented:
Putting a username and password in the URL is indeed deprecated ([RFC 3986 section 3.2.1](https://tools.ietf.org/html/rfc3986#section-3.2.1)). More importantly, support for this was removed in IE7 (apparently before it was deprecated by the IETF) and other browsers have started to [add restrictions](https://medium.com/@lmakarov/say-goodbye-to-urls-with-embedded-credentials-b051f6c7b6a3) on this feature (although I don't think any others have removed support completely).

Given that, I think the right approach here is to have a `demo`-only `/autologin` path (perhaps `/demologin`?) that accepts the username and password as GET-style query parameters. Or we could consider making the regular `/login` endpoint accept its parameters via GET instead of only POST, but I'd prefer making the login-via-GET demo-only (because GET parameters are likely to get logged somewhere)

It would be even better for security if this url contained not the actual password but some sort of consumable login capability, but that feels like overengineering for this use case. (and using the actual password is nice for transparency and consistency with the postgres url variants).

knz commented:
> It would be even better for security if this url contained not the actual password but some sort of consumable login capability

this gives me the idea to (also?) make a URL that takes the auth cookie as a query argument, and injects it as cookie after checking it's valid (i.e. take the query arg and reply to the browser with  `Set-Cookie` header and the same argument).

The `demo` command can pre-create the cookie. This would also be safe(r) for non-demo clusters as auth cookies have an expiry.

knz commented:
> using the actual password is nice for transparency and consistency with the postgres url variants

I feel that as a user I don't care about the consistency between URLs too much. as a user I am not surprised for them to be different from each other. But maybe we an opinion from another routine `demo` user. 

@jseldess  @ericharmeling maybe you can guide our thinking on this?

dhartunian commented:
> Or we could consider making the regular /login endpoint accept its parameters via GET instead of only POST, but I'd prefer making the login-via-GET demo-only (because GET parameters are likely to get logged somewhere)

I like the idea of extending the existing endpoint with new behavior that's only enabled in demo mode. Less chance of missing some important bit of validation or other logic.

> this gives me the idea to (also?) make a URL that takes the auth cookie as a query argument, and injects it as cookie after checking it's valid (i.e. take the query arg and reply to the browser with Set-Cookie header and the same argument).

@knz I like this! It should only be permitted in demo mode. You could even destroy the session and make a new one for same user 🤔 to prevent URL reuse.

knz commented:
> You could even destroy the session and make a new one for same user thinking to prevent URL reuse.

that is clever!

however a tricky bit is that this means the URL printed in the terminal becomes one-time-use. We'd need a new CLI command inside the demo shell to print another login URL. (It's doable so no objection technically.) Would like to hear from a doc / product person for UX feedback though?

dhartunian commented:
> however a tricky bit is that this means the URL printed in the terminal becomes one-time-use. We'd need a new CLI command inside the demo shell to print another login URL if/when needed. (It's doable so no objection technically.) Would like to hear from a doc / product person for UX feedback though?

Could you add a field to the session cookie that marks it as usable from a URL? Then only certain sessions would have this ability, not all of them and only the demo command's session creation would flip the flag.

knz commented:
all right i'll iterate to explore the options

bdarnell commented:
> I feel that as a user I don't care about the consistency between URLs too much

Yeah, it's a very small thing. I'm just thinking that if we put a token other than the password there, people will ask what it is, so I think I'd stick with the password unless we have a good reason.

> You could even destroy the session and make a new one for same user

This is what I meant by a consumable login capability. 

> I like the idea of extending the existing endpoint with new behavior that's only enabled in demo mode. Less chance of missing some important bit of validation or other logic.

But more chance of user confusion. People will look at this demo output and try to script their real clusters based on the same pattern. It can share code with the real login, but I'd much rather have a dedicated demo-only endpoint instead of having demo-only behavior in the "real" endpoint. (unless we decide we want to enable this in non-demo clusters).

jseldess commented:
I'm not sure I have much to add here. I like the idea of making it easier for users to log into the DB console for a demo cluster, but I wonder if auto-login will cause users to miss the fact that there's security at all and that a username/password is required for such cases. Would it be possible to auto-populate the username/password *fields* on the login page but require the user to click **Log in**?

A separate question, would we consider letting users set the username and password they want to use on `cockroach demo`? We could still generate these by default, but it might be nice to provide that option. Similarly, we recently added the ability to specify the port. That helps a lot in tutorials, etc.

awoods187 commented:
I don't think the URL differences will matter to the user. I don't like the idea of the link printed at log-in being a one-time use, I find myself frequently revisiting that link as the easiest way to get to the DB console. I don't think auto-login will cause people to miss security features, when that's the focus we can show them via docs and it will be required outside of the demo. Demo is used to try things and only occasionally will this be security-focused.

knz commented:
As a next step I'll try something with the following properties:

- URL remains multi-use
- make the URLs shorter overall
- demo-only since the token is not consumable

knz committed:
cli/demo,server: new auto-login URL path

Previous to this patch, the user of `cockroach demo` had to perform
two copy-paste operations from their terminal to use the admin UI: one
to copy the URL into their browser, and another oen to copy the
authentication password.

This was inconvenient and thus created friction to exploration and
bottom-up adoption.

This patch rectifies the situation by ensuring that the URL printed
int he terminal allows the user to access the UI directly, in just one
step.

Technically, the approach prototyped here defines a new HTTP GET
endpoint `/demologin` which accepts the user/password parameters as
query arguments, with the same logic (more or less) as `/login`. This
endpoint is reported by (and used only by) `cockroach demo`.

Release note (cli change): The DB console URL printed by `cockroach
demo` now automatically logs the user in when pasted into a web browser.

knz reviewed (commented):
Hi all,
after further research, I have settled to a new HTTP URL `/demologin` which accepts the user/password as URL arguments. As y'all have requested, it takes the credentials directly because we want the URL to remain valid (i.e. not just one use). 

Here is an example such URL:
```
http://127.0.0.1:8080/demologin?password=DGeZilRoXqKHxckh5a4HmyCorkdBd1Ynm0aVRqqLwZP&username=demo
```

However meanwhile, the URL only exists (is only ever defined) from within the `cockroach demo` shell. There is no way to enable it for regular servers.

Would you find this approach satisfactory?

(NB: I know there is another request to also make the URL shorter, in particular by making the *password* shorter for a demo session. This is a legitimate request but is orthogonal, will be addressed by a separate PR.)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56740#-:-MPe7BdaBE3pQwBgg62J:b-ups09m)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aaron-crl and @awoods187)



<!-- Sent from Reviewable.io -->

bdarnell reviewed (approved):
> Reviewed 6 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56740)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aaron-crl and @awoods187)



<!-- Sent from Reviewable.io -->

jseldess reviewed (approved):
LGTM. Thanks, @knz!

knz commented:
TFYRs!

bors r=bdarnell,jseldess
